src/auto_renewal.cairo

```diff
@@ -0,0 +1,242 @@
+                .emit(
+                    Event::toggled_renewal(
+                        toggled_renewal {
+                            domain,
+                            renewer: caller,
+                            limit_price,
+                            is_renewing: new_renew,
+                            last_renewal: new_last_renewal
+                        }
+                    )
+                )
+        }
+
+        fn renew(
+            ref self: ContractState,
+            root_domain: felt252,
+            renewer: ContractAddress,
+            limit_price: u256,
+        ) {
+            self._renew(root_domain, renewer, limit_price);
+        }
+
+        fn batch_renew(
+            ref self: ContractState,
+            domain: array::Array::<felt252>,
+            renewer: array::Array::<starknet::ContractAddress>,
+            limit_price: array::Array::<u256>,
+        ) {
+            assert(domain.len() == renewer.len(), 'Domain & renewer mismatch len');
+            assert(domain.len() == limit_price.len(), 'Domain & price mismatch len');
+
+            let mut domain = domain;
+            let mut renewer = renewer;
+            let mut limit_price = limit_price;
+
+            loop {
+                if domain.len() == 0 {
+                    break;
+                }
+                let _domain = domain.pop_front().expect('pop_front error');
+                let _renewer = renewer.pop_front().expect('pop_front error');
+                let _limit_price = limit_price.pop_front().expect('pop_front error');
+                self._renew(_domain, _renewer, _limit_price);
+            }
+        }
+    }
+
+    //
+    // Internals
+    //
+
+    #[generate_trait]
+    impl InternalImpl of InternalTrait {
+        fn _renew(
+            ref self: ContractState,
+            root_domain: felt252,
+            renewer: ContractAddress,
+            limit_price: u256,
+        ) {
+            let naming = self.naming_contract.read();
+            let can_renew = self._is_renewing.read((renewer, root_domain, limit_price));
+            assert(can_renew == 1, 'Renewal not toggled for domain');
+
+            // Check domain has not been renewed yet this year
+            let block_timestamp: felt252 = get_block_timestamp().into();
+            let last_renewed = self.last_renewal.read((renewer, root_domain));
+            let elapsed: u256 = u256_from_felt252(block_timestamp - last_renewed);
+            let max = u256_from_felt252(86400 * 364);
+            assert(elapsed > max, 'Domain already renewed');
+
+            // Check domain is set to expire within a month
+            let mut domain_arr = ArrayTrait::<felt252>::new();
+            domain_arr.append(root_domain);
+            let expiry = INamingDispatcher {
+                contract_address: naming
+            }.domain_to_expiry(domain_arr);
+            assert(
+                u256_from_felt252(expiry) <= u256_from_felt252(block_timestamp)
```

Comment:
If I understand correctly, the reason why we can't compare felts is because they represent elements from a field with no actual order. Should P-1 be considered bigger or smaller than 0? It's equal to the element -1. In Cairo 0 we used to have various comparison functions working with felts in range [0, 2**128[ or [P-2**128, P[, with the new Cairo we have integer types for this: they only allow values on a specific range on which an order is defined. To me, casting felts to u128 to compare them is the best equivalent to what we used to do before (since the comparison functions would fail outside of u128 range), it should be cheaper than using u256. Here we are comparing timestamps, even u32 would be fine for a few decades but it seems Starkware decided to use u64 as standard, so let's do it with u64.